packages/x-tree-view/src/internals/models/helpers.ts

```diff
@@ -22,33 +22,31 @@ export type OptionalIfEmpty<A extends string, B> = keyof B extends never
     ? Partial<Record<A, B>>
     : Record<A, B>;
 
-export type MergePluginsProperty<
-  TPlugins extends readonly any[],
+export type MergeSignaturesProperty<
+  TSignatures extends readonly any[],
   TProperty extends keyof TreeViewAnyPluginSignature,
-> = TPlugins extends readonly [plugin: infer P, ...otherPlugin: infer R]
+> = TSignatures extends readonly [plugin: infer P, ...otherPlugin: infer R]
   ? P extends TreeViewAnyPluginSignature
-    ? P[TProperty] & MergePluginsProperty<R, TProperty>
+    ? P[TProperty] & MergeSignaturesProperty<R, TProperty>
     : {}
   : {};
 
-export type ConvertPluginsIntoSignatures<TPlugins extends readonly any[]> =
-  TPlugins extends readonly [plugin: infer P, ...otherPlugin: infer R]
-    ? P extends TreeViewPlugin<infer TSignature>
-      ? [TSignature, ...ConvertPluginsIntoSignatures<R>]
-      : ConvertPluginsIntoSignatures<R>
-    : [];
+export type ConvertPluginsIntoSignatures<
+  TPlugins extends readonly TreeViewPlugin<TreeViewAnyPluginSignature>[],
+> = TPlugins extends readonly [plugin: infer TPlugin, ...otherPlugin: infer R]
+  ? R extends readonly TreeViewPlugin<any>[]
+    ? TPlugin extends TreeViewPlugin<infer TSignature>
+      ? readonly [TSignature, ...ConvertPluginsIntoSignatures<R>]
+      : never
+    : never
+  : [];
 
-export interface MergePlugins<TPlugins extends readonly any[]> {
-  state: MergePluginsProperty<TPlugins, 'state'>;
-  instance: MergePluginsProperty<TPlugins, 'instance'>;
-  publicAPI: MergePluginsProperty<TPlugins, 'publicAPI'>;
-  params: MergePluginsProperty<TPlugins, 'params'>;
-  defaultizedParams: MergePluginsProperty<TPlugins, 'defaultizedParams'>;
-  dependantPlugins: MergePluginsProperty<TPlugins, 'dependantPlugins'>;
-  contextValue: MergePluginsProperty<TPlugins, 'contextValue'>;
-  slots: MergePluginsProperty<TPlugins, 'slots'>;
-  slotProps: MergePluginsProperty<TPlugins, 'slotProps'>;
-  events: MergePluginsProperty<TPlugins, 'events'>;
-  models: MergePluginsProperty<TPlugins, 'models'>;
-  experimentalFeatures: MergePluginsProperty<TPlugins, 'experimentalFeatures'>;
-}
+export type ConvertSignaturesIntoPlugins<
+  TSignatures extends readonly TreeViewAnyPluginSignature[],
+> = TSignatures extends readonly [signature: infer TSignature, ...otherSignatures: infer R]
+  ? R extends readonly TreeViewAnyPluginSignature[]
+    ? TSignature extends TreeViewAnyPluginSignature
+      ? readonly [TreeViewPlugin<TSignature>, ...ConvertSignaturesIntoPlugins<R>]
+      : never
+    : never
+  : [];
```

Comment:
that took a while for me to understand :D

I did not know that there is a recursion pattern in TS ... NICE!